pkg/sql/sem/tree/overload.go

```diff
@@ -585,77 +644,84 @@ type typeCheckOverloadState struct {
 
-	var s typeCheckOverloadState
-	s.exprs = exprs
-	s.overloads = overloads
-
 	// Special-case the HomogeneousType overload. We determine its return type by checking that
 	// all parameters have the same type.
-	for i, overload := range overloads {
+	for i := range s.params {
 		// Only one overload can be provided if it has parameters with HomogeneousType.
-		if _, ok := overload.params().(HomogeneousType); ok {
-			if len(overloads) > 1 {
-				return nil, nil, errors.AssertionFailedf(
+		if _, ok := s.params[i].(HomogeneousType); ok {
+			if numOverloads > 1 {
+				return errors.AssertionFailedf(
 					"only one overload can have HomogeneousType parameters")
 			}
-			typedExprs, _, err := TypeCheckSameTypedExprs(ctx, semaCtx, desired, exprs...)
+			typedExprs, _, err := typeCheckSameTypedExprs(ctx, semaCtx, desired, s.exprs...)
 			if err != nil {
-				return nil, nil, err
+				return err
 			}
-			return typedExprs, overloads[i : i+1], nil
+			s.typedExprs = typedExprs
+			s.overloadIdxs = append(s.overloadIdxs[:0], uint8(i))
+			return nil
 		}
 	}
 
 	// Hold the resolved type expressions of the provided exprs, in order.
-	s.typedExprs = make([]TypedExpr, len(exprs))
-	s.constIdxs, s.placeholderIdxs, s.resolvableIdxs = typeCheckSplitExprs(semaCtx, exprs)
+	if cap(s.typedExprs) >= len(s.exprs) {
+		s.typedExprs = s.typedExprs[:len(s.exprs)]
+	} else {
+		s.typedExprs = make([]TypedExpr, len(s.exprs))
+	}
+	s.constIdxs, s.placeholderIdxs, s.resolvableIdxs = typeCheckSplitExprs(
+		semaCtx, s.exprs,
+	)
 
 	// If no overloads are provided, just type check parameters and return.
-	if len(overloads) == 0 {
+	if numOverloads == 0 {
 		for i, ok := s.resolvableIdxs.Next(0); ok; i, ok = s.resolvableIdxs.Next(i + 1) {
-			typ, err := exprs[i].TypeCheck(ctx, semaCtx, types.Any)
+			typ, err := s.exprs[i].TypeCheck(ctx, semaCtx, types.Any)
 			if err != nil {
-				return nil, nil, pgerror.Wrapf(err, pgcode.InvalidParameterValue,
+				return pgerror.Wrapf(err, pgcode.InvalidParameterValue,
 					"error type checking resolved expression:")
 			}
 			s.typedExprs[i] = typ
 		}
-		if err := defaultTypeCheck(ctx, semaCtx, &s, false); err != nil {
-			return nil, nil, err
+		if err := defaultTypeCheck(ctx, semaCtx, s, false); err != nil {
+			return err
 		}
-		return s.typedExprs, nil, nil
+		s.overloadIdxs = s.overloadIdxs[:0]
+		return nil
 	}
 
-	s.overloadIdxs = make([]uint8, len(overloads))
-	for i := 0; i < len(overloads); i++ {
+	if cap(s.overloadIdxs) < numOverloads {
+		s.overloadIdxs = make([]uint8, 0, numOverloads)
+	}
+	s.overloadIdxs = s.overloadIdxs[:numOverloads]
+	for i := range s.overloadIdxs {
 		s.overloadIdxs[i] = uint8(i)
 	}
 
 	// Filter out incorrect parameter length overloads.
-	s.overloadIdxs = filterOverloads(s.overloads, s.overloadIdxs,
-		func(o overloadImpl) bool {
-			return o.params().MatchLen(len(exprs))
-		})
+	exprsLen := len(s.exprs)
+	s.overloadIdxs = filterParams(s.overloadIdxs, s.params, func(
+		params TypeList,
```

Comment:
nit: make `func(params TypeList) bool {` in its own line. just find it easier to read that way.